File: 251051/Birukov/Kafka/publisher/src/main/java/com/publisher/service/TweetService.java
package com.publisher.service;

import com.publisher.dto.request.TweetRequestTo;
import com.publisher.dto.response.TweetResponseTo;
import com.publisher.entity.Tweet;
import com.publisher.entity.Writer;
import com.publisher.mapper.TweetMapper;
import com.publisher.repository.TweetRepository;
import com.publisher.repository.WriterRepository;
import com.publisher.exception.NotFoundException;
import com.publisher.exception.ValidationException;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;

@Service
@Transactional
@Validated
public class TweetService {
	
	@Autowired
	private final TweetRepository tweetRepository;
	
	@Autowired
	private final WriterRepository writerRepository;
	
	@Autowired
	private final TweetMapper tweetMapper;
	
	public TweetService(TweetRepository tweetRepository,
					   WriterRepository writerRepository,
					   TweetMapper tweetMapper) {
		this.tweetRepository = tweetRepository;
		this.writerRepository = writerRepository;
		this.tweetMapper = tweetMapper;
	}
	
	public TweetResponseTo create(@Valid TweetRequestTo request) {
		validateTweetRequest(request);
		
		Writer writer = writerRepository.findById(request.getWriterId())
			.orElseThrow(() -> new NotFoundException("Writer not found: " + request.getWriterId()));
		
		if (tweetRepository.existsByTitle(request.getTitle())) {
			throw new ValidationException("Tweet title already exists: " + request.getTitle());
		}
		
		Tweet tweet = tweetMapper.toEntity(request);
		tweet.setWriter(writer);
		Tweet saved = tweetRepository.save(tweet);
		return tweetMapper.toResponse(saved);
	}
	
	public TweetResponseTo findById(Long id) {
		Tweet tweet = tweetRepository.findById(id)
			.orElseThrow(() -> new NotFoundException("Tweet not found: " + id));
		return tweetMapper.toResponse(tweet);
	}
	
	public List<TweetResponseTo> findAll() {
		return tweetRepository.findAll().stream()
			.map(tweetMapper::toResponse)
			.toList();
	}
	
	public TweetResponseTo update(Long id, @Valid TweetRequestTo request) {
		validateTweetRequest(request);
		
		Tweet existing = tweetRepository.findById(id)
			.orElseThrow(() -> new NotFoundException("Tweet not found: " + id));
		
		Writer writer = writerRepository.findById(request.getWriterId())
			.orElseThrow(() -> new NotFoundException("Writer not found: " + request.getWriterId()));
		
		if (!existing.getTitle().equals(request.getTitle()) && 
			tweetRepository.existsByTitle(request.getTitle())) {
			throw new ValidationException("Tweet title already exists: " + request.getTitle());
		}
		
		existing.setWriter(writer);
		tweetMapper.updateEntity(request, existing);
		Tweet updated = tweetRepository.save(existing);
		return tweetMapper.toResponse(updated);
	}
	
	public void delete(Long id) {
		if (!tweetRepository.existsById(id)) {
			throw new NotFoundException("Tweet not found: " + id);
		}
		tweetRepository.deleteById(id);
	}
	
	public List<TweetResponseTo> findByWriterId(Long writerId) {
		if (!writerRepository.existsById(writerId)) {
			throw new NotFoundException("Writer not found: " + writerId);
		}
		
		return tweetRepository.findByWriterId(writerId).stream()
			.map(tweetMapper::toResponse)
			.toList();
	}
	
	private void validateTweetRequest(TweetRequestTo request) {
		if (request.getWriterId() == null) {
			throw new ValidationException("Writer ID is required");
		}
		if (request.getTitle() == null || request.getTitle().length() < 2 || request.getTitle().length() > 64) {
			throw new ValidationException("Title must be 2-64 characters");
		}
		if (request.getContent() == null || request.getContent().length() < 4 || request.getContent().length() > 2048) {
			throw new ValidationException("Content must be 4-2048 characters");
		}
	}
}

File: 351002/Khmylko/rest-api/src/Service/TweetService.php
<?php
namespace App\Service;

use App\Repository\TweetRepository;
use App\Repository\EditorRepository;
use App\Exception\ApiException;
use App\Exception\ValidationException;
use App\Exception\NotFoundException;

class TweetService {
    private TweetRepository $repository;
    private EditorRepository $editorRepository;

    public function __construct(TweetRepository $repository, EditorRepository $editorRepo) {
        $this->repository = $repository;
        $this->editorRepository = $editorRepo;
    }

    public function getById(int $id): array {
        $tweet = $this->repository->findById($id);
        if (!$tweet) {
            throw new NotFoundException('Tweet', $id);
        }
        return $tweet;
    }

    public function findAll(int $page = 1, int $limit = 10, string $sortBy = 'id', string $order = 'ASC'): array {
        return $this->repository->findAll([], $sortBy, $order, $page, $limit);
    }

    public function create(array $data): array {
        // Конвертируем editorId в editor_id
        if (isset($data['editorId'])) {
            $data['editor_id'] = $data['editorId'];
        }

        $this->validate($data);

        // Проверка существования редактора
        if (!$this->editorRepository->findById($data['editor_id'])) {
            throw new ValidationException("Editor with id {$data['editor_id']} does not exist");
        }

        $tweet = $this->repository->create($data);
        // Конвертируем обратно editor_id в editorId для ответа
        if (isset($tweet['editor_id'])) {
            $tweet['editorId'] = $tweet['editor_id'];
            unset($tweet['editor_id']);
        }

        return $tweet;
    }

    public function update(int $id, array $data): array {
        // Конвертируем editorId в editor_id
        if (isset($data['editorId'])) {
            $data['editor_id'] = $data['editorId'];
        }

        $existing = $this->repository->findById($id);
        if (!$existing) {
            throw new NotFoundException('Tweet', $id);
        }

        if (isset($data['title'])) $existing['title'] = $data['title'];
        if (isset($data['content'])) $existing['content'] = $data['content'];
        if (isset($data['editor_id'])) $existing['editor_id'] = $data['editor_id'];

        $tweet = $this->repository->update($id, $existing);

        // Конвертируем обратно editor_id в editorId для ответа
        if (isset($tweet['editor_id'])) {
            $tweet['editorId'] = $tweet['editor_id'];
            unset($tweet['editor_id']);
        }

        return $tweet;
    }

    public function delete(int $id): void {
        if (!$this->repository->exists($id)) {
            throw new NotFoundException('Tweet', $id);
        }
        $this->repository->delete($id);
    }

    private function validate(array $data): void {
        if (empty($data['title']) || strlen($data['title']) < 2 || strlen($data['title']) > 64) {
            throw new ValidationException("Title must be 2-64 characters");
        }
        if (empty($data['content']) || strlen($data['content']) < 4 || strlen($data['content']) > 2048) {
            throw new ValidationException("Content must be 4-2048 characters");
        }
        if (empty($data['editor_id'])) {
            throw new ValidationException("editorId is required");
        }
    }
}